bin/generate.py

```diff
@@ -1363,24 +1372,34 @@ def parse(key, name, yaml, parent):
                     for child_key, child_yaml in sorted(dictionary.items()):
                         if d.numbered:
                             if is_directory(child_yaml):
-                                testgroup_id += 1
-                                child_name = numbered_testcase_name(
-                                    child_key, testgroup_id, num_testgroups
-                                )
+
+                                def next_testgroup_name():
+                                    nonlocal testgroup_id
+                                    testgroup_id += 1
+                                    return numbered_testcase_name(
+                                        child_key, testgroup_id, num_testgroups
+                                    )
+
+                                child_name = next_testgroup_name
                             elif is_testcase(child_yaml):
-                                testcase_id += 1
-                                child_name = numbered_testcase_name(
-                                    child_key, testcase_id, num_numbered_testcases
-                                )
+
+                                def next_testcase_name():
```

Comment:
Perhaps we can move the `next_testgroup_name` and `next_testcase_name` outside of the loops, to make sure that Python does not re-create them with every iteration :slightly_smiling_face:

Reply:
we need to know child_key?